### ISSUE TITLE
Compile failure: radare2 4.4.0 does not compile on MacOS 10.9 Mavericks

### ISSUE THREAD

**Issue by pineapplemachine** [radareorg/radare2#17264]
I am running MacOS 10.9. Yes, I am aware that it is no longer officially supported, but I am still running it on this machine, and I would like to run the very nice looking rasm2 on it as well.

When running `sys/install.sh` as instructed in the readme (as well as when trying to install via the last supported version of brew), compilation fails because the project uses `clock_gettime` and `CLOCK_MONOTONIC`, which are not supported in MacOS 10.11 and earlier, without using any compatibility layer such as [PosixMachTiming](https://github.com/pineapplemachine/PosixMachTiming/tree/improve-degenerate-case-behavior).

It would be fantastic if such a compatibility layer could be added so that radare2 can be compiled on MacOS 10.9.

As a workaround, though, I was able to get 3.1.3 installed and apparently running fine. I don't suppose anyone can let me know what was the last supported version on MacOS 10.9? I could not find anything in the release changelogs that indicated dropping support for older versions of MacOS.

Thanks!

**Comment by ret2libc** [comment]
Hi! Please do compile the pre-filed template for reporting issues, including the build log and the exact error message you get, otherwise it is hard for us to help you, because not everybody has a MacOS 10.9 to test things.

**Comment by pineapplemachine** [comment]
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | OS X 10.9.5, x86-64
| File format of the file you reverse (mandatory)      | N/A
| Architecture/bits of the file (mandatory)            | N/A
| r2 -v full output, **not truncated** (mandatory)     | N/A

### Expected behavior

The build should complete successfully on MacOS 10.9 Mavericks.

### Actual behavior

The build fails on MacOS 10.9 Mavericks.

### Steps to reproduce the behavior 

Attempt to build radare2 4.4.0 on MacOS 10.11 or earlier. 

This fails because clock_gettime and CLOCK_MONOTONIC were not supported until MacOS 10.12; to make these available on MacOS 10.11 or earlier, a compatibility layer such as [this one](https://github.com/pineapplemachine/PosixMachTiming/tree/improve-degenerate-case-behavior) can be used.

- Please share the binary if it is shareable by drag and dropping it here in a zip archive (mandatory)

N/A

- Use asciinema to describe the issue and share the link here (mandatory if u can't share the binary) 

N/A

- Use code markdown `CODE` to make your code visible

`N/A`

- Or even better, create a Pull-Request containing the test case in the test/ folder. See for example test/new/db/cmd/cmd_search.

N/A

### Additional Logs, screenshots, source-code,  configuration dump, ...

Here is the relevant part of the build log:

``` text
CC run.c
CC r2r.c
clang: warning: argument unused during compilation: '-pie'
clang: warning: argument unused during compilation: '-pie'
clang: warning: argument unused during compilation: '-pie'
run.c:640:3: warning: implicit declaration of function 'clock_gettime' is invalid in C99
      [-Wimplicit-function-declaration]
                clock_gettime (CLOCK_MONOTONIC, &timeout_abs);
                ^
run.c:640:18: error: use of undeclared identifier 'CLOCK_MONOTONIC'
                clock_gettime (CLOCK_MONOTONIC, &timeout_abs);
                               ^
run.c:678:19: error: use of undeclared identifier 'CLOCK_MONOTONIC'
                        clock_gettime (CLOCK_MONOTONIC, &now);
                                       ^
1 warning and 2 errors generated.
make[2]: *** [run.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [all] Error 1
make: *** [all] Error 2
```

**Comment by ret2libc** [comment]
> Attempt to build radare2 4.4.0 on MacOS 10.11 or earlier.

Have you tried to compile it from `master` branch? Latest commit at the moment is 69abbec753f082c71ee1de892c1ff9104422b07f

**Comment by pineapplemachine** [comment]
@ret2libc Cloning the current master branch and attempting to compile produces the same error as before:

``` text
CC run.c
CC r2r.c
clang: warning: argument unused during compilation: '-pie'
clang: warning: argument unused during compilation: '-pie'
clang: warning: argument unused during compilation: '-pie'
run.c:640:3: warning: implicit declaration of function 'clock_gettime' is invalid in
      C99 [-Wimplicit-function-declaration]
                clock_gettime (CLOCK_MONOTONIC, &timeout_abs);
                ^
run.c:640:18: error: use of undeclared identifier 'CLOCK_MONOTONIC'
                clock_gettime (CLOCK_MONOTONIC, &timeout_abs);
                               ^
run.c:678:19: error: use of undeclared identifier 'CLOCK_MONOTONIC'
                        clock_gettime (CLOCK_MONOTONIC, &now);
                                       ^
1 warning and 2 errors generated.
make[2]: *** [run.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [all] Error 1
make: *** [all] Error 2
```

**Comment by ret2libc** [comment]
@pineapplemachine Ok the issue seems to be in r2r, which is our test-runner.

@thestr4ng3r @trufae can we add an option to `configure.acr` to not build r2r? There is one in meson, however meson on MacOS does not sign the radare2 bin, so it is not ideal for a mac user. Avoiding compiling r2r would be the quickest solution(also a valid option by itself) as I guess r2r is mostly run by developers on newer machines.

Ideally, r2r should use some sort of portable function defined in `r_util`, right?

**Comment by trufae** [comment]
Yes. R2r shouldnt reinvent the wheel and use more the rutil and rsocket apis . I complained about this before r2r.c was merged but we priorized the fact to replace r2r.js

**Comment by trufae** [comment]
Imho the easiest solution is to wrap this monotopic code with the same ifdefs as the code in rutil. Also i found a bug in this r2r/run.c code that makes the jq detection fail

**Comment by XVilka** [comment]
Note that "shlr/heap/include/r_jemalloc/internal/jemalloc_internal_defs.h" also has its own `#ifdef` mess. Should be solved as a part of https://github.com/radareorg/radare2/issues/11886

**Comment by XVilka** [comment]
Please try if https://github.com/radareorg/radare2/pull/17275 fixes the problem.

**Comment by thestr4ng3r** [comment]
It won't. macOS before 10.12 has no `clock_gettime()`

**Comment by ret2libc** [comment]
@pineapplemachine as a short term workaround you can do the following:
```
$ ./configure --without-r2r
$ make
$ sudo make install
```
This will not compile r2r, which is anyway an utility used for testing. In the meantime, we are anyway working on a proper fix for this. Thanks, hope it helps!

**Comment by pineapplemachine** [comment]
@ret2libc This resulted in a similar error at a different point in the build process.

I ran `./configure --without-r2r` and then got this error running `make`:

``` text
gcc -c  -MD -fno-common   -fPIC -g -Wall -D__UNIX__=1 -I/Users/pineapple/apps/radare2-4.4.0/libr -I/Users/pineapple/apps/radare2-4.4.0/libr/include -fvisibility=hidden -o rabin2.o rabin2.c
gcc -c  -MD -fno-common   -fPIC -g -Wall -D__UNIX__=1 -I/Users/pineapple/apps/radare2-4.4.0/libr -I/Users/pineapple/apps/radare2-4.4.0/libr/include -fvisibility=hidden -o rafind2.o rafind2.c
gcc -c  -MD -fno-common   -fPIC -g -Wall -D__UNIX__=1 -I/Users/pineapple/apps/radare2-4.4.0/libr -I/Users/pineapple/apps/radare2-4.4.0/libr/include -fvisibility=hidden -o r2agent.o r2agent.c
gcc -c  -MD -fno-common   -fPIC -g -Wall -D__UNIX__=1 -I/Users/pineapple/apps/radare2-4.4.0/libr -I/Users/pineapple/apps/radare2-4.4.0/libr/include -fvisibility=hidden -o radiff2.o radiff2.c
gcc -c  -MD -fno-common   -fPIC -g -Wall -D__UNIX__=1 -I/Users/pineapple/apps/radare2-4.4.0/libr -I/Users/pineapple/apps/radare2-4.4.0/libr/include -fvisibility=hidden -o radare2.o radare2.c
gcc -c  -MD -fno-common   -fPIC -g -Wall -D__UNIX__=1 -I/Users/pineapple/apps/radare2-4.4.0/libr -I/Users/pineapple/apps/radare2-4.4.0/libr/include -fvisibility=hidden -o rahash2.o rahash2.c
"gcc -dynamiclib -o libr_main.dylib -Wl,-install_name,/usr/local/lib/libr_main.dylib main.o rax2.o rasm2.o ragg2.o rarun2.o rabin2.o rafind2.o r2agent.o radiff2.o radare2.o rahash2.o  -L/Users/pineapple/apps/radare2-4.4.0/libr/config -L/Users/pineapple/apps/radare2-4.4.0/libr/cons -L/Users/pineapple/apps/radare2-4.4.0/libr/io -L/Users/pineapple/apps/radare2-4.4.0/libr/util -L/Users/pineapple/apps/radare2-4.4.0/libr/flag -L/Users/pineapple/apps/radare2-4.4.0/libr/asm -L/Users/pineapple/apps/radare2-4.4.0/libr/core -L/Users/pineapple/apps/radare2-4.4.0/libr/debug -L/Users/pineapple/apps/radare2-4.4.0/libr/hash -L/Users/pineapple/apps/radare2-4.4.0/libr/bin -L/Users/pineapple/apps/radare2-4.4.0/libr/lang -L/Users/pineapple/apps/radare2-4.4.0/libr/io -L/Users/pineapple/apps/radare2-4.4.0/libr/anal -L/Users/pineapple/apps/radare2-4.4.0/libr/parse -L/Users/pineapple/apps/radare2-4.4.0/libr/bp -L/Users/pineapple/apps/radare2-4.4.0/libr/egg -L/Users/pineapple/apps/radare2-4.4.0/libr/reg -L/Users/pineapple/apps/radare2-4.4.0/libr/search -L/Users/pineapple/apps/radare2-4.4.0/libr/syscall -L/Users/pineapple/apps/radare2-4.4.0/libr/socket -L/Users/pineapple/apps/radare2-4.4.0/libr/fs -L/Users/pineapple/apps/radare2-4.4.0/libr/magic -L/Users/pineapple/apps/radare2-4.4.0/libr/crypto -lr_config -lr_cons -lr_io -lr_util -lr_flag -lr_asm -lr_core -lr_debug -lr_hash -lr_bin -lr_lang -lr_io -lr_anal -lr_parse -lr_bp -lr_egg -lr_reg -lr_search -lr_syscall -lr_socket -lr_fs -lr_magic -lr_crypto   -fPIC -fvisibility=hidden    -fvisibility=hidden"
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C binr
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all'.
make[2]: Nothing to be done for `all'.
gcc -c  -MD -fno-common   -fPIC -g -Wall -D__UNIX__=1 -I/Users/pineapple/apps/radare2-4.4.0/libr/../shlr/zip/include -pie -I/Users/pineapple/apps/radare2-4.4.0/libr/..//libr/include -o run.o run.c
clang: warning: argument unused during compilation: '-pie'
run.c:640:3: warning: implicit declaration of function 'clock_gettime' is invalid in C99 [-Wimplicit-function-declaration]
                clock_gettime (CLOCK_MONOTONIC, &timeout_abs);
                ^
run.c:640:18: error: use of undeclared identifier 'CLOCK_MONOTONIC'
                clock_gettime (CLOCK_MONOTONIC, &timeout_abs);
                               ^
run.c:678:19: error: use of undeclared identifier 'CLOCK_MONOTONIC'
                        clock_gettime (CLOCK_MONOTONIC, &now);
```

**Comment by ret2libc** [comment]
@pineapplemachine please ensure that you have cleaned your build directory before running that. E.g. 
```
$ git clean -dxf .
$ ./configure --without-r2r
$ make
$ sudo make install
```

File `run.c` should not even be compiled in that case.

**Comment by pineapplemachine** [comment]
Ah, I can see now that in fact the problem was that I didn't realize `--without-r2r` was newly added and just tried to re-build 4.4.0 with it. Sorry about that.

I pulled master and built that with the flag. No build errors and rasm2 seems to be working.

Thanks for your help with this!

**Comment by ret2libc** [comment]
> Thanks for your help with this!

No problem! We are of course working on a proper solution to actually fix r2r in case clock_gettime is not available, but I guess not compiling r2r (since it is not actually a user tool but mainly a development tool) is a good-enough short term solution. Happy it worked out ;)

**Comment by XVilka** [comment]
I cherry-picked the fix into `release-4.5.0` branch for the [future 4.5.1 release](https://github.com/radareorg/radare2/milestone/57).
